xline/src/storage/compact/mod.rs

```diff
@@ -30,60 +31,59 @@ pub(crate) const COMPACT_CHANNEL_SIZE: usize = 32;
 
 /// Compactor trait definition
 #[async_trait]
-pub(crate) trait Compactor: std::fmt::Debug + Send + Sync {
+pub(crate) trait Compactor<C: Compactable>: Send + Sync {
     /// run an auto-compactor
     async fn run(&self);
     /// pause an auto-compactor when the current node denotes to a non-leader role
     fn pause(&self);
     /// resume an auto-compactor when the current becomes a leader
     fn resume(&self);
+    /// Set compactable
+    async fn set_compactable(&self, c: C);
 }
 
 /// `Compactable` trait indicates a method that receives a given revision and proposes a compact proposal
 #[cfg_attr(test, mockall::automock)]
 #[async_trait]
-pub(crate) trait Compactable: std::fmt::Debug + Send + Sync {
+pub(crate) trait Compactable: Send + Sync + 'static {
     /// do compact
-    async fn compact(&self, revision: i64) -> Result<(), ClientError<Command>>;
+    async fn compact(&self, revision: i64) -> Result<(), ExecuteError>;
 }
 
 #[async_trait]
-impl Compactable for Client<Command> {
-    async fn compact(&self, revision: i64) -> Result<(), ClientError<Command>> {
+impl Compactable
+    for Arc<dyn ClientApi<Error = tonic::Status, Cmd = Command> + Sync + Send + 'static>
+{
+    async fn compact(&self, revision: i64) -> Result<(), ExecuteError> {
         let request = CompactionRequest {
             revision,
             physical: false,
         };
         let request_wrapper = RequestWithToken::new_with_token(request.into(), None);
         let cmd = Command::new(vec![], request_wrapper);
-        let _ig = self.propose(cmd, true).await?;
-        Ok(())
+        loop {
+            match self.propose(&cmd, true).await {
+                Ok(res) => return res.map(|_ig| ()),
+                Err(err) => warn!("send compaction request failed, error: {err}"),
+            };
+        }
```

Comment:
Why do we need a loop here? The impl is for a `dyn ClientApi` so the retry logic is a bit confusing here.

Reply:
fixed, now the method signature is changed